File: src/main/java/no/fintlabs/flyt/kafka/instanceflow/consuming/InstanceFlowConsumerRecordMapper.java
package no.fintlabs.flyt.kafka.instanceflow.consuming;

import no.fintlabs.flyt.kafka.instanceflow.headers.InstanceFlowHeadersMapper;
import org.apache.kafka.clients.consumer.ConsumerRecord;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
public class InstanceFlowConsumerRecordMapper {

    private final InstanceFlowHeadersMapper instanceFlowHeadersMapper;

    public InstanceFlowConsumerRecordMapper(InstanceFlowHeadersMapper instanceFlowHeadersMapper) {
        this.instanceFlowHeadersMapper = instanceFlowHeadersMapper;
    }

    public <T> InstanceFlowConsumerRecord<T> toFlytConsumerRecord(ConsumerRecord<String, T> consumerRecord) {
        return new InstanceFlowConsumerRecord<>(instanceFlowHeadersMapper.getInstanceFlowHeaders(consumerRecord.headers()), consumerRecord);
    }

    public <T> List<InstanceFlowConsumerRecord<T>> toFlytConsumerRecords(List<ConsumerRecord<String, T>> consumerRecords) {
        return consumerRecords.stream().map(this::toFlytConsumerRecord).toList();
    }

}

File: src/main/kotlin/no/novari/flyt/kafka/instanceflow/consuming/InstanceFlowConsumerRecordMapper.kt
package no.novari.flyt.kafka.instanceflow.consuming

import no.novari.flyt.kafka.instanceflow.headers.InstanceFlowHeadersMapper
import org.apache.kafka.clients.consumer.ConsumerRecord
import org.springframework.stereotype.Service

@Service
class InstanceFlowConsumerRecordMapper(
    private val instanceFlowHeadersMapper: InstanceFlowHeadersMapper,
) {
    fun <T> toInstanceFlowConsumerRecord(consumerRecord: ConsumerRecord<String, T>): InstanceFlowConsumerRecord<T> =
        InstanceFlowConsumerRecord(
            instanceFlowHeaders = instanceFlowHeadersMapper.getInstanceFlowHeaders(consumerRecord.headers()),
            consumerRecord = consumerRecord,
        )

    fun <T> toInstanceFlowConsumerRecords(
        consumerRecords: List<ConsumerRecord<String, T>>,
    ): List<InstanceFlowConsumerRecord<T>> = consumerRecords.map(::toInstanceFlowConsumerRecord)
}
